src/dbnode/storage/bootstrap/state.go

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) 2020  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package bootstrap
+
+import (
+	"errors"
+
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/persist"
+	"github.com/m3db/m3/src/dbnode/persist/fs"
+	"github.com/m3db/m3/src/x/instrument"
+
+	"go.uber.org/zap"
+)
+
+var (
+	errFilesystemOptsNotSet = errors.New("filesystemOptions not set")
+	errInstrumentOptsNotSet = errors.New("instrumentOptions not set")
+)
+
+// NewState creates state specifically to be used during the bootstrap process.
+// Primarily a mechanism for passing info files along without needing to re-read them at each
+// stage of the bootstrap process.
+func NewState(options StateOptions) (State, error) {
+	if err := options.Validate(); err != nil {
+		return State{}, err
+	}
+	return State{
+		fsOpts:           options.FilesystemOptions(),
+		infoFilesFinders: options.InfoFilesFinders(),
+		iOpts:            options.InstrumentOptions(),
+	}, nil
+}
+
+// InfoFilesForNamespace returns the info files grouped by shard for the provided namespace.
+func (r *State) InfoFilesForNamespace(ns namespace.Metadata) InfoFileResultsPerShard {
+	infoFilesByShard, ok := r.ReadInfoFiles()[ns]
+	// This should never happen as State object is initialized with all namespaces to bootstrap.
```

Comment:
if it should never happen, should this return an `error` instead?

Reply:
+1

Reply:
Yeah, I was using the `EmitAndLogInvariantViolation` method to handle errors since I'd seen that in other parts of the bootstrapping code, but after syncing with @arnikola on Slack about the use case for it, I think it's a bit much. Will switch this to just return a regular error